### PR TITLE
Use partial matching for linter names in exclusion directives

### LINKED ISSUE #872: Remove redundant `_linter` suffix from auto names?
The new `# nolint: xxx.` feature makes it more likely for linter names to appear in source code.
I would therefor suggest we remove the redundant `_linter` suffix from the names to make most lint exclusions look more natural.
Compare:

```
my.bad.name <- function(a, b) { # nolint: object_name_linter.
  # Very long comment explaining why the name came to be and needs to stay # nolint: line_length_linter.
  a + b
}
```

to

```
my.bad.name <- function(a, b) { # nolint: object_name.
  # Very long comment explaining why the name came to be and needs to stay # nolint: line_length.
  a + b
}
```

wdyt?

cc @MichaelChirico @jimhester

### PR DESCRIPTION
**Pull request by AshesITR**
fixes #872 

- transitioned our internal `# nolint`s to the short-hand notation
- we raise a warning if no linter could be matched by an exclusion
- documented the short-hand notation
- added an example in the `using_lintr` vignette
- added tests
- extended NEWS bullet for #660

### CODE REVIEW

File: R/exclude.R [hunk 1]
```diff
@@ -154,13 +157,25 @@ add_excluded_lines <- function(exclusions, excluded_lines, excluded_linters) {
...
   # No match for linter list: Add to global excludes
   if (is.na(linters_string)) {
     exclusions <- add_excluded_lines(exclusions, lines, "")
   } else {
     # Matched a linter list: only add excluded lines for the listed linters.
     excluded_linters <- strsplit(linters_string, exclude_linter_sep)[[1L]]
+    if (!is.null(linter_names)) {
+      idxs <- pmatch(excluded_linters, linter_names, duplicates.ok = TRUE)
+      if (anyNA(idxs)) {
+        bad <- excluded_linters[is.na(idxs)]
```
Comment (MichaelChirico) [inline R/exclude.R:170]: is.na(idxs) is basically run three times here... save & reuse

File: tests/testthat/dummy_projects/project/partially_matched_exclusions.R [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+# nolint start: s. warn (and lint) because of non-unique identifier
+x <- 42; y <- 2
+# nolint end
+
+# nolint start: bogus_linter. warn because of nonexistent identifier
+
+# nolint end
+
+# nolint: hokus_pokus, bogus.
```
Comment (MichaelChirico) [inline tests/testthat/dummy_projects/project/partially_matched_exclusions.R:13]: nit: we spell it hocus pocus :)

File: tests/testthat/test-exclusions.R [hunk 3]
```diff
@@ -68,3 +73,22 @@ test_that("it gives the expected error message when there is mismatch between mu
     fixed = TRUE
   )
 })
+
+test_that("partial matching works for exclusions but warns if no linter found", {
+  read_settings(NULL)
+
+  expect_warning(
```
Comment (MichaelChirico) [inline tests/testthat/test-exclusions.R:80]: is this really the way to do multiple warning tests in testthat? oh boy 😂

the withCallingHandlers approach is not particularly nicer... maybe we need a helper. i've definitely written `expect_warnings()` before.

otoh, it seems like an intentional design decision that testthat has limited matching exactly one warning at a time:

https://testthat.r-lib.org/reference/expect_error.html

**Comment by MichaelChirico** [comment]
LGTM after small fix to reduce redundancy

File: tests/testthat/test-exclusions.R (hunk 3 above)
Comment (AshesITR) [reply tests/testthat/test-exclusions.R:80]: Filed https://github.com/r-lib/testthat/issues/1625 as a suggestion

**Review by MichaelChirico — APPROVED** [review]

**Pull request merged**